src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy

```diff
@@ -454,7 +454,9 @@ public class GenerateProtoTask extends DefaultTask {
       outputDir.mkdirs()
     }
 
-    List<String> dirs = includeDirs*.path.collect { "-I${it}" }
+    // The source directory designated from sourceSet may not actually exist on disk.
+    // "include" it only when it exists, so that Gradle and protoc won't complain.
+    List<String> dirs = includeDirs.filter { it.exists() }*.path.collect { "-I${it}" }
```

Comment:
Also check `isDirectory()`?

Reply:
It's unclear to me whether `ConfigurableFileCollection` adds files recursively or not. Documentation is very vague on this, which is why I didn't use it in the first place. It seems the effective change is moving `exists()` check down to here. Wouldn't it also work if you kept `includeDirs` as a `List<File>`?

Reply:
Since we the issue we are trying to solve is the problem that some directory in the "proto" `SourceDirectorySet` does not exist in configuration phase 

(e.g., 
```
main {
        proto {
            srcDir "${buildDir}/proto"
        }
    }
```
), so we cannot refer to directories of "proto" `SourceDirectorySet` like https://github.com/google/protobuf-gradle-plugin/blob/d25369e20db8a3b3520c3835523641b5e1321ee1/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L317 as Gradle will try to resolve that path and will fail. Instead, we need to use lazy type such as `FileCollection`.

From [`SourceDirectorySet#getSourceDirectories`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/SourceDirectorySet.html#getSourceDirectories--):
> The returned collection is live and reflects changes to this source directory set.

This is the whole point for using `ConfigurableFileCollection `.

I don't think `ConfigurableFileCollection#from` adds files recursively as the [doc](https://docs.gradle.org/current/javadoc/org/gradle/api/file/ConfigurableFileCollection.html#from-java.lang.Object...-) says it just adds paths converted to `ConfigurableFileCollection` by `Project#files` for the given paths and nothing else. `Project#files` only returns a `ConfigurableFileCollection` representation of the given paths and does not expand them.

Reply:
What matters is whether everything under the given directory is treated as the inputs. i.e., anything (under the given directory recursively) changing is treated as the given directory has changed. But `File` and `FileCollection` are just different datatypes for representing a directory. Its name containing "Collection" just means we can represent >= 1 file/directories in a single object but not "create a collection, add each individual files in given directory, return the collection". You can play with `FileCollection#getSingleFile()` or `FileCollection#getFiles()` to confirm this.